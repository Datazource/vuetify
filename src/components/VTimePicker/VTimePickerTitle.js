require('../../stylus/components/_time-picker-title.styl')

import pad from '../VDatePicker/util/pad'

export default {
  name: 'v-time-picker-title',

  props: {
    ampm: Boolean,
    selectingHour: Boolean,
    value: String
  },

  computed: {
    hour () {
      return parseInt(this.value.split(':')[0], 10)
    },
    minute () {
      return parseInt(this.value.split(':')[1], 10)
    },
    period () {
      return this.hour < 12 ? 'am' : 'pm'
    }
  },

  methods: {
    genTime () {
      let hour = this.hour
      if (this.ampm) {
        if (hour === 0) {
          hour = 12
        } else if (this.hour > 12) {
          hour = hour - 12
        }
      }

      return this.$createElement('div', {
        'class': 'time-picker-title__time'
      }, [
        this.$createElement('span', {
          staticClass: 'picker__title__btn time-picker-title__time__hour',
          'class': { active: this.selectingHour },
          on: this.selectingHour ? undefined : {
            click: () => this.$emit('selectingHour', true)
          }
        }, pad(hour)),
        this.$createElement('span', ':'),
        this.$createElement('span', {
          staticClass: 'picker__title__btn time-picker-title__time__minute',
          'class': { active: !this.selectingHour },
          on: this.selectingHour ? {
            click: () => this.$emit('selectingHour', false)
          } : undefined
        }, pad(this.minute))
      ])
    },

    genAMPM () {
      return this.$createElement('div', {
        staticClass: 'time-picker-title__ampm'
      }, [
        this.genPeriod('am'),
        this.genPeriod('pm')
      ])
    },

    genPeriod (period) {
      return this.$createElement('span', {
        staticClass: 'picker__title__btn',
        'class': { active: this.period === period },
        on: this.period === period ? undefined : {
          click: () => this.$emit('period', period)
        }
      }, period)
    }
  },

  render (h) {
    return h('div', {
      staticClass: 'time-picker-title'
    }, [
      this.genTime(),
      this.ampm ? this.genAMPM() : null
    ])
  }
}
