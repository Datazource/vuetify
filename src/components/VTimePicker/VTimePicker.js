// Utils
import pad from '../VDatePicker/util/pad'

// Components
import VPicker from '../VPicker'
import VTimePickerTitle from './VTimePickerTitle'
import VTimePickerClock from './VTimePickerClock'

// Mixins
import Colorable from '../../mixins/colorable'
import Themeable from '../../mixins/themeable'

export default {
  name: 'v-time-picker',

  components: {
    VPicker,
    VTimePickerTitle,
    VTimePickerClock
  },

  mixins: [
    Colorable,
    Themeable
  ],

  data () {
    const { inputHour, inputMinute } = this.getInputTime(this.value)

    return {
      inputHour,
      inputMinute,
      originalHour: inputHour,
      originalMinute: inputMinute,
      selectingHour: true
    }
  },

  props: {
    allowedHours: {
      type: [Array, Object, Function],
      default: () => (null)
    },
    allowedMinutes: {
      type: [Array, Object, Function],
      default: () => (null)
    },
    autosave: Boolean,
    format: {
      type: String,
      default: 'ampm',
      validator (val) {
        return ['ampm', '24hr'].includes(val)
      }
    },
    headerColor: String,
    landscape: Boolean,
    noTitle: Boolean,
    scrollable: Boolean,
    value: null
  },

  computed: {
    hour: {
      get () {
        return this.inputHour == null ? new Date().getHours() : this.inputHour
      },
      set (value) {
        this.inputHour = value
      }
    },
    hour12: {
      get () {
        if (this.hour === 0) {
          return 12
        } else if (this.hour > 12) {
          return this.hour - 12
        } else {
          return this.hour
        }
      },
      set (value) {
        if (this.period === 'pm') {
          this.hour = value === 12 ? value : value + 12
        } else {
          this.hour = value === 12 ? 0 : value
        }
      }
    },
    minute: {
      get () {
        return this.inputMinute == null ? new Date().getMinutes() : this.inputMinute
      },
      set (value) {
        this.inputMinute = value
      }
    },
    period: {
      get () {
        return this.hour < 12 ? 'am' : 'pm'
      },
      set (val) {
        this.hour += val === 'am' ? -12 : 12
      }
    }
  },

  watch: {
    value (value) {
      const { inputHour, inputMinute } = this.getInputTime(value)
      this.inputHour = inputHour
      this.inputMinute = inputMinute
    },
    inputHour (val) {
      this.$emit('input', `${pad(this.hour)}:${pad(this.minute)}`)
    },
    inputMinute (val) {
      this.$emit('input', `${pad(this.hour)}:${pad(this.minute)}`)
    }
  },

  methods: {
    getInputTime (value) {
      let inputHour
      let inputMinute

      if (value instanceof Date) {
        inputHour = value.getHours()
        inputMinute = value.getMinutes()
      } else if (value) {
        const arr = value.replace(/[^:0-9]/, '').split(':')
        inputHour = parseInt(arr[0], 10)
        inputMinute = parseInt(arr[1], 10)

        const [, period] = value.match(/([ap]m)/i) || []
        if (period && period.toLowerCase() === 'pm') {
          inputHour = inputHour === 12 ? inputHour : inputHour + 12
        } else if (period && period.toLowerCase() === 'am') {
          inputHour = inputHour === 12 ? 0 : inputHour
        }
      }

      return { inputHour, inputMinute }
    },
    save () {
      this.originalHour = this.inputHour
      this.originalMinute = this.inputMinute
      this.commit()
    },
    cancel () {
      this.inputHour = this.originalHour
      this.inputMinute = this.originalMinute
      this.commit()
    },
    commit () {
      if (this.$parent && this.$parent.isActive) this.$parent.isActive = false

      // Fix for #1818
      // Wait for data to persist
      // then set selectingHour
      this.$nextTick(() => (this.selectingHour = true))
    },
    genBody () {
      return this.$createElement('v-time-picker-clock', {
        props: {
          allowedValues: this.selectingHour ? this.allowedHours : this.allowedMinutes,
          color: this.color,
          dark: this.dark,
          double: this.selectingHour && this.format !== 'ampm',
          max: this.selectingHour ? (this.format === 'ampm' ? 12 : 23) : 59,
          min: this.selectingHour && this.format === 'ampm' ? 1 : 0,
          pad: this.selectingHour ? 0 : 2,
          rotate: this.selectingHour && this.format === 'ampm' ? 30 : 0,
          scrollable: this.scrollable,
          size: this.landscape ? 250 : 280,
          step: this.selectingHour ? 1 : 5,
          value: this.selectingHour ? (this.format === 'ampm' ? this.hour12 : this.hour) : this.minute
        },
        on: {
          input: value => {
            if (!this.selectingHour) {
              this.minute = value
            } else if (this.format === '24hr') {
              this.hour = value
            } else {
              this.hour12 = value
            }
          },
          change: () => {
            if (!this.selectingHour && this.autosave) {
              this.save()
            }

            if (this.selectingHour) {
              this.selectingHour = !this.selectingHour
            }
          }
        },
        ref: 'clock'
      })
    },

    genTitle () {
      return this.$createElement('v-time-picker-title', {
        props: {
          ampm: this.format === 'ampm',
          value: pad(this.hour) + ':' + pad(this.minute),
          selectingHour: this.selectingHour
        },
        on: {
          selectingHour: value => (this.selectingHour = value),
          period: value => (this.period = value)
        },
        ref: 'title',
        slot: 'title'
      })
    }
  },

  render (h) {
    const children = [
      this.noTitle ? null : this.genTitle(),
      h('div', {
        style: {
          width: '100%',
          height: '100%'
        },
        key: this.selectingHour
      }, [this.genBody()]),
      this.$scopedSlots.default ? h('<div>', {
        slot: 'actions'
      }, [this.$scopedSlots.default({
        save: this.save,
        cancel: this.cancel
      })]) : null
    ]

    const data = {
      staticClass: 'picker--time',
      props: {
        landscape: this.landscape,
        dark: this.dark,
        light: this.light,
        color: this.headerColor
      }
    }

    return h('v-picker', data, children)
  }
}
