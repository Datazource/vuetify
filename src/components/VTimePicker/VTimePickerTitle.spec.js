import { VTimePickerTitle } from '~components/VTimePicker'
import { test } from '~util/testing'

test('VTimePickerTitle.js', ({ mount }) => {
  it('should render component in 24hr', () => {
    const wrapper = mount(VTimePickerTitle, {
      propsData: {
        value: '14:13',
        ampm: false
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component in 12hr', () => {
    const wrapper = mount(VTimePickerTitle, {
      propsData: {
        value: '14:13',
        ampm: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component when selecting hour', () => {
    const wrapper = mount(VTimePickerTitle, {
      propsData: {
        value: '14:13',
        selectingHour: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should emit event when clicked on am/pm', () => {
    const wrapper = mount(VTimePickerTitle, {
      propsData: {
        value: '14:13',
        ampm: true
      }
    })

    const period = jest.fn()
    wrapper.vm.$on('period', period)

    wrapper.find('.time-picker-title__ampm span.active')[0].trigger('click')
    expect(period).not.toBeCalled()
    wrapper.find('.time-picker-title__ampm span:not(.active)')[0].trigger('click')
    expect(period).toBeCalledWith('am')
  })

  it('should emit event when clicked on hours/minutes', async () => {
    const wrapper = mount(VTimePickerTitle, {
      propsData: {
        value: '14:13'
      }
    })

    const selectingHour = jest.fn()
    wrapper.vm.$on('selectingHour', selectingHour)

    wrapper.find('.time-picker-title__time__minute')[0].trigger('click')
    expect(selectingHour).not.toBeCalled()
    wrapper.find('.time-picker-title__time__hour')[0].trigger('click')
    expect(selectingHour).toBeCalledWith(true)
    wrapper.setProps({ selectingHour: true })
    await wrapper.vm.$nextTick()
    wrapper.find('.time-picker-title__time__minute')[0].trigger('click')
    expect(selectingHour).toBeCalledWith(false)
  })
})
