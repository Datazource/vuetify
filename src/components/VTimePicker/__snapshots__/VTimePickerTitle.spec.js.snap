// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTimePickerTitle.js should render component in 12hr 1`] = `

<div class="time-picker-title">
  <div class="time-picker-title__time">
    <span class="picker__title__btn time-picker-title__time__hour">
      02
    </span>
    <span>
      :
    </span>
    <span class="picker__title__btn time-picker-title__time__minute active">
      13
    </span>
  </div>
  <div class="time-picker-title__ampm">
    <span class="picker__title__btn">
      am
    </span>
    <span class="picker__title__btn active">
      pm
    </span>
  </div>
</div>

`;

exports[`VTimePickerTitle.js should render component in 24hr 1`] = `

<div class="time-picker-title">
  <div class="time-picker-title__time">
    <span class="picker__title__btn time-picker-title__time__hour">
      14
    </span>
    <span>
      :
    </span>
    <span class="picker__title__btn time-picker-title__time__minute active">
      13
    </span>
  </div>
</div>

`;

exports[`VTimePickerTitle.js should render component when selecting hour 1`] = `

<div class="time-picker-title">
  <div class="time-picker-title__time">
    <span class="picker__title__btn time-picker-title__time__hour active">
      14
    </span>
    <span>
      :
    </span>
    <span class="picker__title__btn time-picker-title__time__minute">
      13
    </span>
  </div>
</div>

`;
